### ISSUE TITLE
Question: could asyncify be used to emulate pthread?

### ISSUE THREAD

**Issue by allsey87** [GoogleChromeLabs/asyncify#6]
Would some sort of cooperative multitasking would be possible with Asyncify? For example, when a yield point or sleep is reached the stack is unwound, saved, and another stack for the same WebAssembly module is resumed?

It seems like this could be nice solution to threading (without web workers) until WebAssembly has native support for threads.

**Comment by RReverser** [comment]
In theory it could, although, of course, it wouldn't provide performance benefits of multithreading, so it's a bit moot. But why would you use that instead of real WebAssembly threads, e.g. via Emscripten pthread support?

**Comment by allsey87** [comment]
Well, I am working on a project where there will be several WebAssembly modules running in parallel, each with its own threads. I am currently writing a runtime to support this and would have preferred to use WASI as much as possible to keep things portable. However, from what I have read, there probably won't be any support for threads in WASI until WebAssembly has native thread support (which I suspect will not land for a couple years). Moreover Emscripten's `STANDALONE` mode does not support threads, so I have to give up on WASI for now.

> e.g. via Emscripten pthread support?

This is what I will end up doing with the runtime based on what Emscripten generates with the `MINIMAL_RUNTIME` option.

**Comment by RReverser** [comment]
Right, you still need a custom syscall to spawn a thread in the first place. However, if you just want to run individual modules in parallel, then you don't really need pthread support in Wasm - just use the host threading capabilities to spawn each module in its own thread. In that case Wasm won't need any syscalls as it won't even know it's running in a separate thread.

Wasm threading is only useful when you need to share memory between instances of the same module. If you can avoid that, things get a lot easier.

Either way, I don't think it's very relevant to asyncify - as replied above, yes, it's possible, but you have little to gain with this approach.

**Comment by allsey87** [comment]
> However, if you just want to run individual modules in parallel, then you don't really need pthread support in Wasm

The modules themselves use threading internally, so I will need to create a custom runtime that shares N web workers between the threads from M modules. But this is no longer relevant to Asyncify. Thanks for all the write ups and documentation @RReverser it has been invaluable to my project!
